unikernel.ml

```diff
@@ -530,7 +568,189 @@ struct
+            | Ok csrf ->
+                Lwt.return
+                  (reply reqd ~content_type:"text/html"
+                     (Dashboard.dashboard_layout user
+                        ~page_title:"Account | Mollymawk"
+                        ~content:
+                          (User_account.user_account_layout ~csrf user cookie
+                             now)
+                        ~icon:"/images/robur.png" ())
+                     ~header_list:[ ("X-MOLLY-CSRF", csrf) ]
+                     `OK)
+            | Error err ->
+                Lwt.return
+                  (reply reqd ~content_type:"text/html"
+                     (Dashboard.dashboard_layout user
+                        ~page_title:"500 | Mollymawk"
+                        ~content:(Error_page.error_layout err)
+                        ~icon:"/images/robur.png" ())
+                     `Internal_server_error))
+        | None ->
+            let error =
+              {
+                Utils.Status.code = 401;
+                title = "Unauthenticated";
+                success = false;
+                data = "Auth cookie not found";
+              }
+            in
+            Lwt.return
+              (reply reqd ~content_type:"text/html"
+                 (Dashboard.dashboard_layout user ~page_title:"401 | Mollymawk"
+                    ~content:(Error_page.error_layout error)
+                    ~icon:"/images/robur.png" ())
+                 `Unauthorized))
+    | Error (`Msg err) ->
+        let error =
+          {
+            Utils.Status.code = 401;
+            title = "Unauthenticated";
+            success = false;
+            data = err;
+          }
+        in
+        Lwt.return
+          (reply reqd ~content_type:"text/html"
+             (Dashboard.dashboard_layout user ~page_title:"401 | Mollymawk"
+                ~content:(Error_page.error_layout error)
+                ~icon:"/images/robur.png" ())
+             `Unauthorized)
+
+  let update_password json store reqd (user : User_model.user) =
+    match json with
+    | `Assoc xs -> (
+        match
+          Utils.Json.
+            ( get "current_password" xs,
+              get "new_password" xs,
+              get "confirm_password" xs )
+        with
+        | ( Some (`String current_password),
+            Some (`String new_password),
+            Some (`String confirm_password) ) -> (
+            let now = Ptime.v (P.now_d_ps ()) in
+            let new_password_hash =
+              User_model.hash_password ~password:new_password ~uuid:user.uuid
+            in
+            match
+              ( String.equal user.password
+                  (* TODO: remove concats from current_password and fix all json parsing, remove Utils.clean_string for good*)
+                  (* Changing passwords will break login because of Utils.clean_string*)
+                  (User_model.hash_password
+                     ~password:("\"" ^ current_password ^ "\"")
```

Comment:
@hannesm This is why reset_password was breaking. 
So using Json `member` always wraps the string in quotes `"some text"` and this was why I initially wrote the `Utils.Json.clean_string` thingy. 

So for reset password, I am not using `member` and so the passwords were read as-is without any quotes attached and this is why the hash couldn't match with what was in storage. 

So manually adding the quotes fixes that..which is a bit sad. 

So I wonder if we can just get rid of all the instances of `YoJson.member` in this PR, because currently if you reset a password and try to login, the hash won't match because login still uses the old `yojson.member` method.

and if we update login to not use `Yojson.member` and merge this PR we "may" not be able to login to our accounts

Reply:
After merging, to login add quotes to your password. so if it was previously `password` now on the login form enter `"password"` and you can use the update password under account to update the password to `password` without the qoutes and it should work given the hashes are different.

Reply:
done in 1499eac